vault/identity_mfa.go

```diff
@@ -0,0 +1,1676 @@
+	}
+
+	if mConfigRaw == nil {
+		return nil, nil
+	}
+
+	mConfig, ok := mConfigRaw.(*mfa.Config)
+	if !ok {
+		return nil, fmt.Errorf("failed to declare the type of fetched MFA config")
+	}
+
+	if clone {
+		return mConfig.Clone()
+	}
+
+	return mConfig, nil
+}
+
+// TODO: should I remove this? seems not used anywhere
+func (b *SystemBackend) MemDBMFAConfigByID(mConfigID string, clone bool) (*mfa.Config, error) {
+	if mConfigID == "" {
+		return nil, fmt.Errorf("missing config id")
+	}
+
+	txn := b.db.Txn(false)
+
+	return b.MemDBMFAConfigByIDInTxn(txn, mConfigID, clone)
+}
+
+func (b *SystemBackend) deleteMFAConfigByMethodName(ctx context.Context, methodName string) error {
+	var err error
+
+	if methodName == "" {
+		return fmt.Errorf("missing method name")
+	}
+
+	b.mfaLock.Lock()
+	defer b.mfaLock.Unlock()
+
+	// Delete the config from storage
+	entryIndex := mfaConfigPrefix + methodName
+	err = b.Core.systemBarrierView.Delete(ctx, entryIndex)
+	if err != nil {
+		return err
+	}
+
+	// Create a MemDB transaction to delete config
+	txn := b.db.Txn(true)
+	defer txn.Abort()
+
+	mConfig, err := b.MemDBMFAConfigByMethodNameInTxn(txn, methodName, false)
+	if err != nil {
+		return err
+	}
+
+	if mConfig == nil {
+		return nil
+	}
+
+	if mConfig.Type == "totp" && mConfig.ID != "" {
+		// This is best effort; if they end up hanging around it's okay, they're encrypted anyways
+		if err := logical.ClearView(ctx, NewBarrierView(b.Core.barrier, fmt.Sprintf("%s%s", mfaTOTPKeysPrefix, mConfig.ID))); err != nil {
+			b.mfaLogger.Warn("unable to clear TOTP keys", "method", mConfig.Name, "error", err)
+		}
+	}
+
+	// Delete the config from MemDB
+	err = b.MemDBDeleteMFAConfigByMethodNameInTxn(txn, methodName)
+	if err != nil {
+		return err
+	}
+
+	txn.Commit()
+
+	return nil
+}
+
+func (b *SystemBackend) MemDBDeleteMFAConfigByMethodNameInTxn(txn *memdb.Txn, methodName string) error {
+	if methodName == "" {
+		return nil
+	}
+
+	if txn == nil {
+		return fmt.Errorf("txn is nil")
+	}
+
+	mConfig, err := b.MemDBMFAConfigByMethodNameInTxn(txn, methodName, false)
+	if err != nil {
+		return err
+	}
+
+	if mConfig == nil {
+		return nil
+	}
+
+	err = txn.Delete(memDBMFAConfigsTable, mConfig)
+	if err != nil {
+		return errwrap.Wrapf("failed to delete MFA config from memdb: {{err}}", err)
+	}
+
+	return nil
+}
+
+// TODO: should this be remove as it is not used anywhere?
+func (b *SystemBackend) MemDBDeleteMFAConfigByIDInTxn(txn *memdb.Txn, configID string) error {
+	if configID == "" {
+		return nil
+	}
+
+	if txn == nil {
+		return fmt.Errorf("txn is nil")
+	}
+
+	mConfig, err := b.MemDBMFAConfigByIDInTxn(txn, configID, false)
+	if err != nil {
+		return err
+	}
+
+	if mConfig == nil {
+		return nil
+	}
+
+	err = txn.Delete(memDBMFAConfigsTable, mConfig)
+	if err != nil {
+		return errwrap.Wrapf("failed to delete MFA config from memdb: {{err}}", err)
+	}
+
+	return nil
+}
+
+func (b *SystemBackend) putMFAConfig(ctx context.Context, mConfig *mfa.Config) error {
```

Comment:
I have a question about the following stated in the MFA RFC:
MFA methods configured via identity/mfa/method are not stored within the identity store's storage view, as that would make it awkward for sharing them across nested namespaces.  Instead the methods are stored as a single table record indexed by namespace ID.
So, based on the `putMFAConfig` method, the configuration is stored in the systemBarrierView under `mfa/method/<mConfig.Name>`. So, I am wondering if what I am reading is correct or not, and if not where in the code I can find how the config is stored under the identity view? 
The second part of the requirement in the RFC is clear as the table record needs to be indexed by namespaceID.